google/cloud/ndb/_cache.py

```diff
@@ -583,20 +586,28 @@ def future_info(self, key, value):
 
 
 @tasklets.tasklet
-def global_lock_for_read(key):
+def global_lock_for_read(key, prev_value):
     """Lock a key for a read (lookup) operation by setting a special value.
 
     Lock may be preempted by a parallel write (put) operation.
 
     Args:
         key (bytes): The key to lock.
+        prev_value (bytes): The cache value previously read from the global cache.
+            Should be either :data:`None` or `_LOCKED_FOR_WRITE` if a key was written
+            recently.
 
     Returns:
         tasklets.Future: Eventual result will be lock value (``bytes``) written to
             Datastore for the given key, or :data:`None` if the lock was not acquired.
     """
     lock = _LOCKED_FOR_READ + str(uuid.uuid4()).encode("ascii")
-    lock_acquired = yield global_set_if_not_exists(key, lock, expires=_LOCK_TIME)
+    if prev_value == _LOCKED_FOR_WRITE:
+        yield global_watch(key, prev_value)
+        lock_acquired = yield global_compare_and_swap(key, lock, expires=_LOCK_TIME)
+    else:
+        lock_acquired = yield global_set_if_not_exists(key, lock, expires=_LOCK_TIME)
+
```

Comment:
I wonder if this can be simplified by expanding update-value a tad to take an initial old value and then using that here.

Reply:
If you're referring to `_update_key`, that gets called from the lock/unlock methods for write locks, which are called from `_datastore_api.put`, which doesn't do a `global_get`, so we haven't already gotten the value yet when that gets called.

Reply:
Sorry, yes `_update_key`.  `_update_key` would have to change to accept an initial value.  While this is only a few lines of code, there is a lot of context to keep track of.  Not repeating it, and adding some comments explaining why we use a watch/CAS in one case and set-if-not-empty in the other would be beneficial, IMO.

This is just a suggestion. Feel free to ignore. :)

Reply:
Oh, I see. I was confused because we aren't even using that here, but you're suggesting that may be we should...

Reply:
After looking at it for a minute, I think I'm going to leave it alone.